state/metadata/state_manager.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package metadata
+
+import (
+	"bytes"
+
+	"github.com/ava-labs/hypersdk/chain"
+)
+
+const (
+	defaultHeightPrefix    byte = 0x0
+	defaultTimestampPrefix byte = 0x1
+	defaultFeePrefix       byte = 0x2
+
+	DefaultMinimumPrefix byte = 0x3
+)
+
+var _ chain.MetadataManager = (*MetadataManager)(nil)
+
+type MetadataManager struct {
+	heightPrefix    []byte
+	feePrefix       []byte
+	timestampPrefix []byte
+}
+
+func NewManager(
+	heightPrefix []byte,
+	feePrefix []byte,
+	timestampPrefix []byte,
+) MetadataManager {
+	return MetadataManager{
+		heightPrefix:    heightPrefix,
+		feePrefix:       feePrefix,
+		timestampPrefix: timestampPrefix,
+	}
+}
+
+func NewDefaultManager() MetadataManager {
+	return MetadataManager{
+		heightPrefix:    []byte{defaultHeightPrefix},
+		feePrefix:       []byte{defaultFeePrefix},
+		timestampPrefix: []byte{defaultTimestampPrefix},
+	}
+}
+
+func (m MetadataManager) FeePrefix() []byte {
+	return m.feePrefix
+}
+
+func (m MetadataManager) HeightPrefix() []byte {
+	return m.heightPrefix
+}
+
+func (m MetadataManager) TimestampPrefix() []byte {
+	return m.timestampPrefix
+}
+
+// Returns true if all prefixes in `m` and `vmPrefixes` are unique
+func HasConflictingPrefixes(
+	m chain.MetadataManager,
+	vmPrefixes [][]byte,
+) bool {
+	prefixes := [][]byte{
+		m.HeightPrefix(),
+		m.FeePrefix(),
+		m.TimestampPrefix(),
+	}
+
+	prefixes = append(prefixes, vmPrefixes...)
+	verifiedPrefixes := make([][]byte, 0)
+
+	for _, p := range prefixes {
+		for _, vp := range verifiedPrefixes {
+			if bytes.HasPrefix(p, vp) || bytes.HasPrefix(vp, p) {
+				return true
+			}
+		}
+
+		verifiedPrefixes = append(verifiedPrefixes, p)
+	}
```

Comment:
hmm - not necessary for this PR since performance is not important here. Could we make this check faster?

Reply:
Could you clarify? I'm a bit confused between the statements "performance is not important here" and "can we make this check faster?"

Reply:
This is called once on startup, so performance is not important here and therefore not needed for this PR. What I meant by "could we make this check faster" is, is there a way to eliminate duplicate work here?

I thought we could do something simpler like include the index in the first loop and check against all later indices. If we are at index i > j, then there's no need to check index i against j again because we already did. This is already what we're doing though because `verifiedPrefixes` starts out as empty, so rather than checking the first prefix against all the others and then decrease the number to check against from there, we check the first prefix against nothing and then the next prefix against one and so on.